Source/EmbeddedBoundary/WarpXInitEB.cpp

```diff
@@ -291,9 +291,215 @@ WarpX::ScaleAreas (ablastr::fields::VectorField& face_areas,
     }
 }
 
+void
+WarpX::MarkUpdateCellsStairCase (
+    std::array< std::unique_ptr<amrex::iMultiFab>,3> & eb_update,
+    ablastr::fields::VectorField const& field,
+    amrex::EBFArrayBoxFactory const & eb_fact )
+{
+
+    using ablastr::fields::Direction;
+    using warpx::fields::FieldType;
+
+    // Extract structures for embedded boundaries
+    amrex::FabArray<amrex::EBCellFlagFab> const& eb_flag = eb_fact.getMultiEBCellFlagFab();
+
+    for (int idim = 0; idim < 3; ++idim) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for (amrex::MFIter mfi(*field[idim]); mfi.isValid(); ++mfi) {
+
+            const amrex::Box& box = mfi.tilebox();
+            amrex::Array4<int> const & eb_update_arr = eb_update[idim]->array(mfi);
+
+            // Check if the box (including one layer of guard cells) contains a mix of covered and regular cells
+            const amrex::Box& eb_info_box = mfi.tilebox(amrex::IntVect::TheCellVector()).grow(1);
+            amrex::FabType const fab_type = eb_flag[mfi].getType( eb_info_box );
+
+            if (fab_type == amrex::FabType::regular) { // All cells in the box are regular
+
+                // Every cell in box is all regular: update field in every cell
+                amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    eb_update_arr(i, j, k) = 1;
+                });
+
+            } else if (fab_type == amrex::FabType::covered) { // All cells in the box are covered
+
+                // Every cell in box is all covered: do not update field
+                amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    eb_update_arr(i, j, k) = 0;
+                });
+
+            } else { // The box contains a mix of covered and regular cells
+
+                auto const & flag = eb_flag[mfi].array();
+                auto index_type = field[idim]->ixType();
+
+                amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+
+                    // Stair-case approximation: If neighboring cells are either partially
+                    // or fully covered: do not update field
```

Comment:
If I'm understanding the logic here correctly, `eb_flag` is defined on the cell centers and returns `false` for `isRegular()` if any part of the cell is touched by the EB. The reason you have to check the cell below the current one when dealing with a nodal MF is because for the same index, nodal MFs are offset by half a cell downward from the cell centered MF. I think it might be worth adding a bit more explanation in the comment to make it easier for future readers to understand why it is only necessary to check the index offset by -1 (but never +1).

Reply:
Yes, that's correct. I added more comments ; could you check if that's clearer now?